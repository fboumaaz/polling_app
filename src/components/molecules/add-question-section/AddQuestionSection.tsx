import React, { FC, useEffect, useState } from "react";
import {
  TextField, Button, Typography,
  Grid, Box, Paper
} from "@mui/material";
import ClearSharpIcon from "@mui/icons-material/ClearSharp";
import { v4 as uuid } from "uuid";

import AddQuestionSectionStyle from "./AddQuestionSectionStyle";
import optionsProps from "../../shared/interfaces/OptionsProps";

interface Props {
  question: string;
  setQuestion: (value: string) => void;
  options: Array<optionsProps>;
  setOptions: (optionsProps: Array<optionsProps>) => void;
};

const AddQestionSection: FC<Props> = ({
                                        question,
                                        options,
                                        setOptions,
                                        setQuestion
                                      }: Props): JSX.Element => {
  const [currentAnswer, setCurrentAnswer] = useState("");
  const [disabled, setDisabled] = useState(!question);
  const [disableButton, setDisableButton] = useState(!currentAnswer);

  const addQuestion = (e: EventTarget & (HTMLTextAreaElement | HTMLInputElement)): void => {
    if (e.value.length >= 80) e.disabled = true;
    else if (e.disabled) e.disabled = false;
    setQuestion(e.value);
  };
  const submitOption = (): void => setOptions([...options, {
    id: uuid(),
    answer: currentAnswer,
    vote: 0
  }]);

  const addOption = (e: EventTarget & (HTMLTextAreaElement | HTMLInputElement)): void => {
    if (e.value.length >= 80) e.disabled = true;
    else if (e.disabled) e.disabled = false;
    setCurrentAnswer(e.value);
  };

  const editOption = (e: EventTarget & (HTMLTextAreaElement | HTMLInputElement)): void => {
    options.map((option) => {
      if (option.id === e.id) option.answer = e.value;
    });
    if (e.value.length >= 80) e.disabled = true;
    else if (e.disabled) e.disabled = false;
    setOptions([...options]);
  };

  const removeOption = (id: string): void => {
    const newOption = options.filter((option) => option.id !== id);
    setOptions([...newOption]);
  };

  const restOptions = (): void => {
    setQuestion('');
    setOptions([]);
  };

  const OptionsList = (
    options.map((option, index) => (
      <div className="input-group" id="input-group" key={index.toString()}>
        <TextField
          label={`Answer-${index + 1}`}
          variant="standard"
          key={`field ${option.id}`}
          id={`${option.id}`}
          defaultValue={option.answer}
          className="text-field"
          inputProps={{ maxLength: 80 }}
          onChange={(e) => editOption(e.target)}
        />
        <Button
          color="error"
          variant="outlined"
          className="input-button"
          sx={{ mt: 3, ml: 2 }}
          onClick={() => removeOption(option.id)}
          size="small"
          key={`button ${index}`}
        >
          <ClearSharpIcon sx={{ fontSize: "medium" }} />
        </Button>
      </div>
    ))
  );

  useEffect(() => {
    setCurrentAnswer("");
    if (options.length >= 10 || !question) {
      setDisabled(true);
    } else if (disabled && question) setDisabled(false);
  }, [options, question]);

  useEffect(() => {
    if (!currentAnswer) {
      setDisableButton(true);
    } else if (disableButton && question && currentAnswer) setDisableButton(false);
  }, [currentAnswer]);

  return (
    <AddQuestionSectionStyle>
      <Paper variant="outlined" sx={{ flexGrow: 1, my: 2, p: 1.5}}>
        <Grid container rowSpacing={8}>
          <Grid item xs={12}>
            <Typography
              variant="h5"
              color="secondary"
              align="center"
              sx={{ mb: 4 }}
              gutterBottom
              key="title"
            >
              Question / Answers form
            </Typography>
            <TextField
              key="question_key"
              label="What's your question?"
              value={question}
              onChange={(e) => addQuestion(e.target)}
              variant="standard"
              inputProps={{ maxLength: 80 }}
              className="text-field"
              id="question_key"
            />
            {OptionsList}
            <div className="input-group">
              <TextField
                label="Add new answer"
                variant="standard"
                key="new_option_key"
                value={currentAnswer}
                disabled={disabled}
                inputProps={{ maxLength: 80 }}
                onChange={(e) => addOption(e.target)}
                className="text-field"
                id="add_new_option"
              />
              <Button
                color="primary"
                sx={{ mt: 2.4, ml: 2, p: 0.3 }}
                className="input-button"
                disabled={disableButton}
                onClick={() => submitOption()}
                variant="contained"
                component="span"
                size="small"
              >
                Add
              </Button>
            </div>
          </Grid>
          <Grid item xs={6}>
            <Box display="flex" justifyContent="start-end">
              <Typography
                component="p"
                variant="subtitle2"
                sx={{ ml: 1, p: 0.3, mt: 1}}
              >
                {`${options.length} / 10 possible answers`}
              </Typography>
            </Box>
          </Grid>
          <Grid item xs={6}>
            <Box display="flex" justifyContent="flex-end">
              <Button
                color="secondary"
                variant="contained"
                sx={{ mb: 1, p: 0.3, mr: 1 }}
                onClick={restOptions}
              >
                Rest
              </Button>
            </Box>
          </Grid>
        </Grid>
      </Paper>
    </AddQuestionSectionStyle>
  );
};

export default AddQestionSection;
