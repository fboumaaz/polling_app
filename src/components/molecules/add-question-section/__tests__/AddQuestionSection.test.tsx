import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import AddQuestionSection from "../AddQuestionSection";
import { Button, Typography } from "@mui/material";
import optionsData from "../../../shared/tests/options.data";

configure({ adapter: new Adapter() });

describe("AddQuestionSection component", () => {
  const setQuestion = jest.fn();
  const setOptions = jest.fn();
  describe("content", () => {
    const wrapper = shallow(
      <AddQuestionSection
        options={[]} question={""}
        setQuestion={setQuestion} setOptions={setOptions}
      />);

    it("has a title", () => {
      expect(wrapper.find(Typography).at(0).text()).toBe("Question / Answers form");
    });

    it("has a TextField for question", () => {
      expect(wrapper.find("#question_key")).toHaveLength(1);
      expect(wrapper.find("#question_key").prop("label")).toBe("What's your question?");
    });

    it("call setQuestion when changing TextField question", () => {
      expect(setQuestion).not.toHaveBeenCalled();
      wrapper.find("#question_key").simulate("change", { target: { value: "what?" } });
      expect(setQuestion).toHaveBeenCalled();
    });

    it("has a reset button", () => {
      expect(wrapper.find(Button).at(1).text()).toBe("Rest");
    });

    it("renders snapshots", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("when options and question ar not filled yet", () => {
    const wrapper = shallow(
      <AddQuestionSection
        options={[]} question={""}
        setQuestion={setQuestion} setOptions={setOptions}
      />);

    it("has a add input disabled", () => {
      expect(wrapper.find("#add_new_option")).toHaveLength(1);
      expect(wrapper.find("#add_new_option").prop("disabled")).toBe(true);
    });
  });

  describe("when question is filled", () => {
    const wrapper = shallow(
      <AddQuestionSection
        options={[]} question={"what is the capital of the Netherlands?"}
        setQuestion={setQuestion} setOptions={setOptions}
      />);

    it("has a add input not disabled", () => {
      expect(wrapper.find("#add_new_option")).toHaveLength(1);
      expect(wrapper.find("#add_new_option").prop("disabled")).toBe(false);
    });
  });

  describe("when question && answer are filled", () => {
    const wrapper = shallow(
      <AddQuestionSection
        options={optionsData} question={"what is the capital of the Netherlands?"}
        setQuestion={() => jest.fn()} setOptions={setOptions}
      />);

    it("lists existing options", () => {
      expect(wrapper.find("#input-group")).toHaveLength(3);
    });

    it("calls setOptions when reset button is clicked", () => {
      expect(setOptions).not.toHaveBeenCalled();
      wrapper.find(Button).at(3).simulate("click");
      expect(setOptions).toHaveBeenCalled();
    });
  });
});
