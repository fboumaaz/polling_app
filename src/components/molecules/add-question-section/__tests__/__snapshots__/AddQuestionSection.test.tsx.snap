// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuestionSection component content renders snapshots 1`] = `ShallowWrapper {}`;

exports[`AddQuestionSection component renders snapshots 1`] = `ShallowWrapper {}`;

exports[`AddQuestionSection component when options and question ar not filled yet renders snapshots 1`] = `ShallowWrapper {}`;

exports[`AddQuestionSection component when question && answer are filled renders snapshots 1`] = `ShallowWrapper {}`;

exports[`AddQuestionSection component when question is filled renders snapshots 1`] = `ShallowWrapper {}`;
